# HW2--Python Essentials + NumPy

Exercise #1 (10 points)

In the classic card game War, a standard deck of (52) playing cards is shuffled and distributed evenly among two players (26 cards each). For each round of gameplay, the players lay out one card at a time (for our purposes, in order of how they have been dealt) and the player with the higher card wins the round (where Ace = 1, 2-10 equal their face value, and Jack, Queen, and King = 11, 12, 13, respectively, across each of the four suits: diamonds, hearts, clubs, spades). Ties are broken with additional plays until one player wins the full round (i.e., all of the cards played during the round, including those that were tied). Although variations of the game can continue until one player wins all of the cards, let's consider the simple case in which cards won during a previous round cannot be replayed.

A variant of War, called Addition War, consists of two cards being played by each player at a time, where the player with the higher sum of two card values wins the round. Similar to the original game, ties are broken with additional plays of two cards each until one player wins all of the cards in the round. Once all cards have been played (once), the player with the higher score (sum of collected card values) wins the game. If the players run out of cards in the middle of a tiebreaker, assume that each player's respective cards (for that round) are returned to them for the final score.


Similar to the deal function, this function does not take in any arguments. The function should print information about each round of play, which will assist in validating the operation of the function. Finally, the function should store and output the initial value totals for each player's hand (i.e., after dealing), the final value totals for each player (i.e., after playing the full game), and the winner (i.e., the player with the higher total). You may choose the details of how all of this information is reported and stored, but the output should be clear and well organized.

Validate the operation of your function by simulating one game without ties and one game with ties. You may need to simulate several games to generate both outcomes.

### a. (1 point) Write (and execute) a statement that will generate a list of values associated with a standard deck of playing cards. You do not need to differentiate between suits (i.e., diamonds, hearts, clubs, spades), as they have no relevance for this particular game. Also report the total value associated with a full deck of cards, which represents the maximum total for a game.

In [1]:
import numpy as np
L = np.array([range(1,14)])
# Concatenation by row
deck = np.concatenate([L, L], axis=0)
deck = np.concatenate([deck, deck], axis=0)
print(deck)
print('Total value: ', sum(sum(deck)))

[[ 1  2  3  4  5  6  7  8  9 10 11 12 13]
 [ 1  2  3  4  5  6  7  8  9 10 11 12 13]
 [ 1  2  3  4  5  6  7  8  9 10 11 12 13]
 [ 1  2  3  4  5  6  7  8  9 10 11 12 13]]
Total value:  364


### b. (2 points) Define a function deal that will shuffle and distribute the 52 playing cards evenly to two players (26 each) and return a tuple of each player's hand (as a list of values). The function does not need to take in any arguments, and should create the deck of values internally (i.e., you should not need to input the deck of values into the function; you may reuse the statement you developed for part a). You may assume that you are always playing with a standard deck (52 cards). You may make use of any NumPy functionality to perform the shuffle operation. 

In [2]:
# define the deal function
def deal():
    flat_deck = deck.flatten()                           ## flatten the 4-dimension array into one-dimension
    np.random.shuffle(flat_deck)                         ## shuffle 52 items in flatten array
    player_1 = flat_deck[0:26]                           ## the first half of cards are assigned to player 1
    player_2 = flat_deck[26:]                            ## rest of cards are assigned to player 2
    return (list(player_1),list(player_2))

### Validate the operation of your function by dealing a hand to two players and displaying each player's hand along with their number of cards and the associated total value.

In [3]:
deal()
hand = deal()
print('Player 1: ', hand[0])
print('Player 2: ', hand[1])
print('Card number of Player 1: ', len(hand[0]))
print('Card number of Player 2: ', len(hand[1]))
print('Total value for Player 1: ', sum(hand[0]))
print('Total value for Player 2: ', sum(hand[1]))

Player 1:  [9, 7, 10, 5, 2, 3, 7, 6, 11, 2, 11, 13, 3, 7, 9, 5, 12, 11, 8, 1, 5, 10, 3, 8, 4, 1]
Player 2:  [4, 9, 4, 2, 11, 5, 13, 1, 9, 1, 8, 3, 4, 6, 12, 8, 6, 2, 6, 10, 7, 13, 10, 13, 12, 12]
Card number of Player 1:  26
Card number of Player 2:  26
Total value for Player 1:  173
Total value for Player 2:  191


### c. (4 points) Define a function play that will simulate a single game of Addition War between two players. Note that you will need to account for rounds in which a given player wins outright (i.e., after the initial play of two cards) and for rounds in which one or more ties occur. You will also need to terminate the game appropriately (i.e., after each player has played all of their cards once).

In [4]:
def play():
    # shuffle and deal cards
    flat_deck = deck.flatten()                       
    np.random.shuffle(flat_deck)                      
    player_1 = flat_deck[0:26]                           
    player_2 = flat_deck[26:]
    total_p1 = []
    total_p2 = []
    tie = []
    # simulate rounds in which a given player wins outright one or more ties occur
    for i in range(1,14):
        if sum(player_1[(i*2)-2:(i*2)]) > sum(player_2[(i*2)-2:(i*2)]):    ## player 1 wins
            tie.append(0)
            score_p1 = sum(player_1[(i*2)-2:(i*2)]) + sum(player_2[(i*2)-2:(i*2)]) + tie[(i-2)]
            score_p2 = 0
            print('Round ' + str(i) + ': '+ 'Player 1 win')
            print('Player 1 card: ', player_1[(i*2)-2:(i*2)])
            print('Player 2 card: ', player_2[(i*2)-2:(i*2)])
            print('Player 1 score: ', score_p1)
            print('Player 2 score: ', score_p2)
            total_p1.append(score_p1)  
            total_p2.append(score_p2)
        elif sum(player_1[(i*2)-2:(i*2)]) < sum(player_2[(i*2)-2:(i*2)]):  ## player 2 wins
            tie.append(0)
            score_p1 = 0
            score_p2 = sum(player_1[(i*2)-2:(i*2)]) + sum(player_2[(i*2)-2:(i*2)]) + tie[(i-2)]
            print('Round ' + str(i) + ': '+ 'Player 2 win')
            print('Player 1 card: ', player_1[(i*2)-2:(i*2)])
            print('Player 2 card: ', player_2[(i*2)-2:(i*2)])
            print('Player 1 score: ', score_p1)
            print('Player 2 score: ', score_p2)
            total_p1.append(score_p1)  
            total_p2.append(score_p2)
        else:               ## if tie happens
            if i == 1:      ### if happens in first round: there is no accumulative tie score
                print('Round ' + str(i) + ': ' + 'Tie')
                print('Player 1 card: ', player_1[(i*2)-2:(i*2)])
                print('Player 2 card: ', player_2[(i*2)-2:(i*2)])
                tie.append(sum(player_1[(i*2)-2:(i*2)]) + sum(player_2[(i*2)-2:(i*2)]))
                print('Tie score: ', tie[(i-1)])
                print('Player 1 score: ', 0)
                print('Player 2 score: ', 0)
            elif i == 13:   ### if happens in last round: I should split tie score evenly for two players
                score_p1 = sum(player_1[(i*2)-2:(i*2)]) + (tie[(i-2)])/2
                score_p2 = sum(player_2[(i*2)-2:(i*2)]) + (tie[(i-2)])/2
                print('Round ' + str(i) + ': '+ 'Tie')
                print('Player 1 card: ', player_1[(i*2)-2:(i*2)])
                print('Player 2 card: ', player_2[(i*2)-2:(i*2)])
                print('Player 1 score: ', score_p1)
                print('Player 2 score: ', score_p2)
                total_p1.append(score_p1)  
                total_p2.append(score_p2)
            else:           ### if happens in between: a consecutive tie might happen, so I calculate accumulative tie score               
                print('Round ' + str(i) + ': ' + 'Tie')
                print('Player 1 card: ', player_1[(i*2)-2:(i*2)])
                print('Player 2 card: ', player_2[(i*2)-2:(i*2)])
                tie.append(sum(player_1[(i*2)-2:(i*2)]) + sum(player_2[(i*2)-2:(i*2)]) + tie[i-2])
                print('Tie score: ', tie[(i-1)])
                print('Player 1 score: ', 0)
                print('Player 2 score: ', 0)
                
    # sum up players' scores in every round
    print('Player 1 total score: ', sum(total_p1))
    print('Player 2 total score: ', sum(total_p2))
    # decide the final winner by comparing the total score 
    if sum(total_p2)>sum(total_p1):
        print('Player 2 win!')
    elif sum(total_p2)<sum(total_p1):
        print('Player 1 win!')
    else: print('Tie!')

play()  ## display result

Round 1: Player 1 win
Player 1 card:  [10  8]
Player 2 card:  [9 3]
Player 1 score:  30
Player 2 score:  0
Round 2: Player 2 win
Player 1 card:  [6 7]
Player 2 card:  [12  5]
Player 1 score:  0
Player 2 score:  30
Round 3: Player 2 win
Player 1 card:  [11 10]
Player 2 card:  [ 9 13]
Player 1 score:  0
Player 2 score:  43
Round 4: Player 2 win
Player 1 card:  [1 2]
Player 2 card:  [6 9]
Player 1 score:  0
Player 2 score:  18
Round 5: Player 1 win
Player 1 card:  [ 5 13]
Player 2 card:  [ 1 13]
Player 1 score:  32
Player 2 score:  0
Round 6: Player 1 win
Player 1 card:  [11  5]
Player 2 card:  [9 1]
Player 1 score:  26
Player 2 score:  0
Round 7: Player 2 win
Player 1 card:  [6 4]
Player 2 card:  [7 4]
Player 1 score:  0
Player 2 score:  21
Round 8: Player 2 win
Player 1 card:  [3 2]
Player 2 card:  [12  4]
Player 1 score:  0
Player 2 score:  21
Round 9: Player 2 win
Player 1 card:  [2 3]
Player 2 card:  [11  2]
Player 1 score:  0
Player 2 score:  18
Round 10: Player 2 win
Player 1 card:

### d. (3 points) Given your implemented game, let's perform some analysis of the gameplay. For brevity in your notebook, you may want to create a copy of your play function that does not include any (active) print statements to use for this part of the assignment. Simulate n = 20 games of Addition War, and provide the following results and analysis (1 point each):

In [5]:
# the majority of play_1() function is similar to the previous function: play()
def play_1(n=1):
    """This function will simulate n times (n =20) the game and return the values including: initial scores of 
    each player(player1_initial, player2_initial), final scores of each player(total_p1, total_p2) and whether 
    player 1 or player 2 win the game(p1_win and p2_win, 1 represents win, 0 represents loss)"""
    results = []        # store each game result in a list
    for i in range(0,n):     ## fairly same as play() function
        flat_deck = deck.flatten()                       
        np.random.shuffle(flat_deck)                      
        player_1 = flat_deck[0:26]                           
        player_2 = flat_deck[26:]
        player1_initial = sum(player_1)
        player2_initial = sum(player_2)
        total_p1 = []
        total_p2 = []
        tie = []
        p1_win = 0
        p2_win = 0
        for i in range(1,14):
            if sum(player_1[(i*2)-2:(i*2)]) > sum(player_2[(i*2)-2:(i*2)]):    ## player 1 wins
                tie.append(0)
                score_p1 = sum(player_1[(i*2)-2:(i*2)]) + sum(player_2[(i*2)-2:(i*2)]) + tie[(i-2)]
                score_p2 = 0
                total_p1.append(score_p1)  
                total_p2.append(score_p2)
            elif sum(player_1[(i*2)-2:(i*2)]) < sum(player_2[(i*2)-2:(i*2)]):  ## player 2 wins
                tie.append(0)
                score_p1 = 0
                score_p2 = sum(player_1[(i*2)-2:(i*2)]) + sum(player_2[(i*2)-2:(i*2)]) + tie[(i-2)]
                total_p1.append(score_p1)  
                total_p2.append(score_p2)
            else:               ## if tie happens
                if i == 1:      ### if happens in first round: there is no accumulative tie score
                    tie.append(sum(player_1[(i*2)-2:(i*2)]) + sum(player_2[(i*2)-2:(i*2)]))
                elif i == 13:   ### if happens in last round: I should split tie score evenly for two players
                    score_p1 = sum(player_1[(i*2)-2:(i*2)]) + (tie[(i-2)])/2
                    score_p2 = sum(player_2[(i*2)-2:(i*2)]) + (tie[(i-2)])/2
                    total_p1.append(score_p1)  
                    total_p2.append(score_p2)
                else:           ### if happens in between, a consecutive tie might happen, so I calculate accumulative tie score
                    tie.append(sum(player_1[(i*2)-2:(i*2)]) + sum(player_2[(i*2)-2:(i*2)]) + tie[i-2])
        # summrize who wins each game
        if sum(total_p2)>sum(total_p1):    ## if player 2 wins
            p2_win += 1
            p1_win += 0
        elif sum(total_p2)<sum(total_p1):  ## if player 1 wins
            p1_win += 1
            p2_win += 0
        else:                              ## if tie
            p1_win += 0
            p2_win += 0
        # extract values that I want
        result = (player1_initial, player2_initial, sum(total_p1), sum(total_p2), p1_win, p2_win)
        # bring all result together
        results.append(result)
        # show the results in terms of an array
        my_array = np.array(results)
    return my_array

In [6]:
game_results = play_1(20)
print('Results for 20 games: ', game_results)

Results for 20 games:  [[179. 185. 183. 181.   1.   0.]
 [182. 182. 139. 225.   0.   1.]
 [170. 194. 181. 183.   0.   1.]
 [162. 202. 176. 188.   0.   1.]
 [158. 206. 100. 264.   0.   1.]
 [173. 191. 144. 220.   0.   1.]
 [195. 169. 225. 139.   1.   0.]
 [180. 184. 188. 176.   1.   0.]
 [195. 169. 231. 133.   1.   0.]
 [195. 169. 230. 134.   1.   0.]
 [192. 172. 285.  79.   1.   0.]
 [182. 182. 238. 126.   1.   0.]
 [197. 167. 214. 150.   1.   0.]
 [199. 165. 290.  74.   1.   0.]
 [204. 160. 260. 104.   1.   0.]
 [191. 173. 177. 187.   0.   1.]
 [174. 190. 105. 259.   0.   1.]
 [173. 191. 169. 195.   0.   1.]
 [182. 182. 199. 165.   1.   0.]
 [187. 177. 161. 203.   0.   1.]]


#### Extract the final scores for the winning and losing players

In [7]:
# extract winner and losser's scores in each game   
# extract winner's final scores
p1_win_final = game_results[game_results[:,-2]==1,2]
p2_win_final = game_results[game_results[:,-1]==1,3]
win_final = np.concatenate([p1_win_final,p2_win_final])
# extract losser's final scores
p1_loss_final = game_results[game_results[:,-1]==1,2]
p2_loss_final = game_results[game_results[:,-2]==1,3]
loss_final = np.concatenate([p1_loss_final,p2_loss_final])
print('Winner scores are: ', win_final)
print('Losser scores are: ', loss_final)

Winner scores are:  [183. 225. 188. 231. 230. 285. 238. 214. 290. 260. 199. 225. 183. 188.
 264. 220. 187. 259. 195. 203.]
Losser scores are:  [139. 181. 176. 100. 144. 177. 105. 169. 161. 181. 139. 176. 133. 134.
  79. 126. 150.  74. 104. 165.]


#### Produce descriptive statistics for each scenario, specifically the minimum, mean, and maximum totals after each game.

In order to find more steady descriptive statistics for this question, I tried to run play_1(n) fucntion with a much higher n value, such as 1000 and 10000. What I found is, the min score for losser could be 0 and highest score for winner could be 364 if playing many times. And the maximum score for losser could be 181 and minimum for winner could be 183, which means that the game result is possible to be decided by only one point. In addition, Winner's mean score is around 60% of total mark (364).

In [8]:
print('Maximum score for winner: ', max(win_final))
print('Maximum score for losser: ', max(loss_final))
print('Minimum score for winner: ', min(win_final))
print('Minimum score for losser: ', min(loss_final))
print('Mean score for winner: ', np.mean(win_final))
print('Mean score for losser: ', np.mean(loss_final))

Maximum score for winner:  290.0
Maximum score for losser:  181.0
Minimum score for winner:  183.0
Minimum score for losser:  74.0
Mean score for winner:  223.35
Mean score for losser:  140.65


#### What proportion of games resulted in a winner that had a lower initial total than their opponent? 

I tried many times for this question and found that the rate is basically between 0.15 to 0.2. Also, I ran 'play_1(n = 1000)', and the result (approximately 0.19) by doing so was consistent with my previous guess. So I would like to say that a player with a lower initial has quite small possibility (around 20%) to win the game.

In [9]:
# extraxt the games player 1 win
player1_win = game_results[game_results[:,-2]==1, :]
# extraxt the games player 2 win
player2_win = game_results[game_results[:,-1]==1, :]
# from the games player 1 win, extract games win with lower initial than player 2
player1_win_low_ini = player1_win[player1_win[:,0]<player1_win[:,1],:]
# from the games player 2 win, extract games win with lower initial than player 1
player2_win_low_ini = player2_win[player2_win[:,0]>player2_win[:,1],:]
# sum up the number of game that player wins the game but with lower initial score and calculate the proportion
game_low_ini = len(player1_win_low_ini) + len(player2_win_low_ini)
print('Proportion of games a winner that had a lower initial total than their opponent: ', game_low_ini/20)

Proportion of games a winner that had a lower initial total than their opponent:  0.2


#### What was the lowest initial total to win a game (or conversely, the highest initial total to lose a game)?

Also, I tried n=1000 and n=10000 to find a more precise answer for this question, the result is, based on the data I got, the initial to win a game should not be lower than 40% of the totoal point (364), which is around 150 and the highest initial should be around 60% of 364, which is around 215.

In [10]:
# extra the min initial score when win and loss
print('Lowest initial total to win a game: ', min(min(game_results[game_results[:,-2]==1,0]),min(game_results[game_results[:,-1]==1,1])))
print('Highest initial total to loss a game: ', max(max(game_results[game_results[:,-2]==1,1]),max(game_results[game_results[:,-1]==1,0])))

Lowest initial total to win a game:  173.0
Highest initial total to loss a game:  191.0


#### Calculate the (linear) correlation between the initial and final totals for winners vs. losers? What do you observe about the respective correlations, and what do these correlations tell you about the relationship between the initial and final totals?

I simulated the game 10000 times to find a steady answer for this question, and the final correlation rate should be around 60%. Another finding which is quite easy to notice is the correlation coefficients of initial and final scores are same for Losser and Winner. Therefore, initial and final scores are positively correlated and for both winner and losser in each game. In other word, the higher the initial score, the higher final score, and vice versa.

In [12]:
# extract winner's initial and final scores
p1_win_init = game_results[game_results[:,-2]==1,0]
p2_win_init = game_results[game_results[:,-1]==1,1]
win_init = np.concatenate([p1_win_init,p2_win_init])
print("Winner's initial score: ", win_init)
# I have extracted win_final in previous question
print("Winner's final score: ", win_final)

# extract losser's initial and final scores
p1_loss_init = game_results[game_results[:,-1]==1,0]
p2_loss_init = game_results[game_results[:,-2]==1,1]
loss_init = np.concatenate([p1_loss_init,p2_loss_init])
print("Losser's initial score: ", loss_init)
# I have extracted win_final in previous question
print("Losser's final score: ", loss_final)

# calculate the correlation coefficient
print("Winner's Correlation Coefficient: ", np.corrcoef(win_init,win_final)[0][1])
print("Losser's Correlation Coefficient: ", np.corrcoef(loss_init,loss_final)[0][1])

Winner's initial score:  [179. 195. 180. 195. 195. 192. 182. 197. 199. 204. 182. 182. 194. 202.
 206. 191. 173. 190. 191. 177.]
Winner's final score:  [183. 225. 188. 231. 230. 285. 238. 214. 290. 260. 199. 225. 183. 188.
 264. 220. 187. 259. 195. 203.]
Losser's initial score:  [182. 170. 162. 158. 173. 191. 174. 173. 187. 185. 169. 184. 169. 169.
 172. 182. 167. 165. 160. 182.]
Losser's final score:  [139. 181. 176. 100. 144. 177. 105. 169. 161. 181. 139. 176. 133. 134.
  79. 126. 150.  74. 104. 165.]
Winner's Correlation Coefficient:  0.4983135998743003
Losser's Correlation Coefficient:  0.4983135998743004
